pallets/ctype/src/lib.rs

```diff
@@ -26,73 +26,148 @@ pub mod benchmarking;
 #[cfg(test)]
 mod tests;
 
+#[cfg(test)]
+mod mock;
+
 pub mod default_weights;
 pub use default_weights::WeightInfo;
 
-use frame_support::{decl_error, decl_event, decl_module, decl_storage, dispatch::DispatchResult, ensure, StorageMap};
+use codec::{Decode, Encode};
+use frame_support::ensure;
 use frame_system::{self, ensure_signed};
+use sp_std::fmt::Debug;
+
+pub use pallet::*;
+
+/// An operation to create a new CTYPE.
+/// The struct implements the DidOperation trait, and as such it must contain
+/// information about the creator's DID, the type of DID key required to
+/// verify the operation signature, and the tx counter to protect against replay
+/// attacks. The struct has the following fields:
+/// * creator_did: the DID of the CTYPE creator
+/// * hash: the CTYPE hash
+/// * tx_counter: the DID tx counter to mitigate replay attacks
+#[derive(Clone, Decode, Encode, PartialEq)]
+pub struct CtypeCreationOperation<T: Config> {
+	creator_did: <T as did::Config>::DidIdentifier,
+	hash: <T as frame_system::Config>::Hash,
```

Comment:
If you pass both of them separately you can derive debug. Makes it also more generic since the requirements are not as tight as requiring the whole `Config`.

```suggestion
pub struct CtypeCreationOperation<DidIdentifier, Hash> {
	creator_did: DidIdentifier,
	hash: Hash,
```

Reply:
This is the way I was going initially. But then when I wanted to use the DID pallet structs in the CTYPE pallet, I was having weird type casting issues that I did not manage to solve in an afternoon (something like `type annotations needed: cannot satisfy <_ as did::Config>::DidIdentifier == <T as did::Config>::DidIdentifier).

Plus, I think it would be quite repetitive to have the same type constraints, which are basically a logic of the DID pallet, in all the other pallets that want to implement the `DidOperation` trait. In this way, a pallet just must require the implementation of the did Config Trait as well, and can be readily used.

Reply:
Since we have the did pallet everywhere you can just use these associated types from there. No need to repeat them anywhere.

Reply:
Yes but to use the pallet associated type, we need a reference to the pallet's Config trait, since we cannot simply reference it from everywhere. That is why the type parametrisation over the pallet's Config trait, which in turn extends the did pallet's Config trait. Are there other ways to access the did pallet's `DidIdentifier` type directly that I am not aware of?

Reply:
Doesn't `<Config as did::Config>::DidIdentifier` work?

Reply:
got the problem now! as we discussed, keep it like it is for now.